File: Client/DotNet/Thriot.Client.DotNet/Platform/OccasionallyConnectionClient.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Json;
using System.Text;

namespace Thriot.Client.DotNet.Platform
{
    /// <summary>
    /// Represents an occasionally connected client. There is no persistent connection, every operation is a standalone REST API request.
    /// </summary>
    public class OccasionallyConnectionClient
    {
        private readonly IRestConnection _restConnection;

        /// <summary>
        /// Create new instance of an occasionally connected client
        /// </summary>
        /// <param name="baseUrl">Base API url</param>
        /// <param name="deviceId">Unique device identifier</param>
        /// <param name="apiKey">Api key of the device, enclosing network or enclosing service</param>
        public OccasionallyConnectionClient(string baseUrl, string deviceId, string apiKey)
            : this(baseUrl, deviceId, apiKey, new RestConnection())
        {
        }

        private OccasionallyConnectionClient(string baseUrl, string deviceId, string apiKey, IRestConnection restConnection)
        {
            _restConnection = restConnection;
            _restConnection.Setup(baseUrl, new Dictionary<string, string>
            {
                {"X-DeviceId", deviceId},
                {"X-ApiKey", apiKey}
            });
        }

        /// <summary>
        /// Record telemetry data.
        /// 
        /// Send POST request to the APIROOT/telemetry Url
        /// </summary>
        /// <param name="message">Telemetry data in JSON format with maximum length of 1024 characters</param>
        /// <exception cref="WebException">This exception indicates some service level error. Please refer to the HTTP error code for more information</exception>
        public void RecordTelmetryData(string message)
        {
            _restConnection.Post("telemetry", message);
        }

        /// <summary>
        /// Send message to a device. Only to devices in the same network can messages be sent.
        /// 
        /// Send POST request to the APIROOT/message/deviceId Url
        /// </summary>
        /// <param name="deviceId">Target device id</param>
        /// <param name="message">Message with maximum length of 512 bytes</param>
        /// <exception cref="WebException">This exception indicates some service level error. Please refer to the HTTP error code for more information</exception>
        public void SendMessageTo(string deviceId, string message)
        {
            string payload = "\"" + Convert.ToBase64String(Encoding.UTF8.GetBytes(message)) + "\"";

            _restConnection.Post("messages/sendto/" + deviceId, payload);
        }

        
        /// <summary>
        /// Receives a message sent to the current device. After receiving the message will be removed from the queue. In case of any error the message cannot be rereceived.
        /// This method implements QoS 0-like reliability from the reveiver's perspective.
        /// 
        /// Send GET request to the APIROOT/message/forget Url
        /// </summary>
        /// <returns></returns>
        /// <exception cref="WebException">This exception indicates some service level error. Please refer to the HTTP error code for more information</exception>
        public PushedMessage ReceiveAndForgetMessage()
        {
            var result = _restConnection.Get("messages/forget");

            if (result == "null")
                return null;

            return ConvertToPushedMessage(result);
        }

        /// <summary>
        /// Receives a message sent to the current device. After receiving you should call the <see cref="CommitMessage"/> method that will remove the message from the queue. 
        /// In case of any error the message can be rereceived until the <see cref="CommitMessage"/> is called.
        /// This method implements QoS 1-like reliability from the reveiver's perspective.
        ///  
        /// Send GET request to the APIROOT/message/peek Url
        /// </summary>
        /// <returns></returns>
        /// <exception cref="WebException">This exception indicates some service level error. Please refer to the HTTP error code for more information</exception>
        public PushedMessage PeekMessage()
        {
            var result = _restConnection.Get("messages/peek");

            if (result == "null")
                return null;

            return ConvertToPushedMessage(result);
        }

        /// <summary>
        /// Informs the service that the message received by the <see cref="PeekMessage"/> method was successfully processed.
        /// This method implements QoS 1-like reliability from the reveiver's perspective.
        /// 
        /// Send POST request to the APIROOT/message/commit Url
        /// </summary>
        /// <exception cref="WebException">This exception indicates some service level error. Please refer to the HTTP error code for more information</exception>
        public void CommitMessage()
        {
            _restConnection.Post("messages/commit", "");
        }

        private static PushedMessage ConvertToPushedMessage(string outgoingMessageStr)
        {
            var jsonSerializer = new DataContractJsonSerializer(typeof (PushedMessageDto));

            using (var ms = new MemoryStream(Encoding.UTF8.GetBytes(outgoingMessageStr)))
            {
                ms.Position = 0;
                var outgoingMessage = (PushedMessageDto) jsonSerializer.ReadObject(ms);

                return new PushedMessage(outgoingMessage.MessageId, outgoingMessage.Timestamp, outgoingMessage.Payload, outgoingMessage.SenderDeviceId);
            }
        }

        [DataContract]
        public class PushedMessageDto
        {
            [DataMember]
            public string Payload { get; set; }

            [DataMember]
            public long Timestamp { get; set; }

            [DataMember]
            public int MessageId { get; set; }

            [DataMember]
            public string SenderDeviceId { get; set; }
        }
    }
}

File: Client/Linux/lib/Platform/OccasionallyConnectionClient.cpp
#include "PlatformClient.h"
#include "ArduinoJson.h"
#include "../RestConnection.h"
#include "../base64.h"

namespace Thriot { namespace Platform {

/**
Create a new instance of the occasionally connection client

@param baseUrl Base Platform API url
@param deviceId Unique 32-characters long id
@param apiKey device key, enclosing network key or session key

*/
OccasionallyConnectionClient::OccasionallyConnectionClient(const string& baseUrl, const string& deviceId, const string& apiKey)
{
	_restConnection = new RestConnection(baseUrl);
	_restConnection->AddRequestHeader("X-DeviceId", deviceId);
	_restConnection->AddRequestHeader("X-ApiKey", apiKey);
}

// In this class:
// we are not handling character codes above or eq 0x80
// TODO: fix to support at least latin1 chars

/**
Record telemetry data. Send telemetry message to the service.
On error returns the http status code, otherwise 0.

@param message Telemetry data to be recorded. Maximum 1024 characters.

@return Status code (0 if everything went well) */
int OccasionallyConnectionClient::RecordTelemetryData(const string& message)
{
	Response httpResponse = _restConnection->Post("telemetry", "application/json", message);

	if(httpResponse.Code!= 200)
		return httpResponse.Code;

	return 0;
}

/**
Send message to an other device from the actually signed in device.
On error returns the http status code, otherwise 0.

@param deviceId Target device id
@param message Message to be sent (maximum 512 charactres)

@return Status code*/
int OccasionallyConnectionClient::SendMessageTo(const string& deviceId, const string& message)
{
	string payload = "\"" + base64_encode((unsigned char *)message.c_str(), message.size()) + "\"";

	Response httpResponse = _restConnection->Post("messages/sendto/" + deviceId, "application/json", payload);

	if(httpResponse.Code!= 204)
		return httpResponse.Code;

	return 0;
}

/**
Receive message and forget it on the service side.
If an error occures the message cannot be rereceived. This is QoS 0-level reliability from the receiver's side.
On error returns an uninitialized message.

@return Message instance*/
PushedMessage OccasionallyConnectionClient::ReceiveAndForgetMessage()
{
	return Receive("messages/forget");
}

/**
Receive message and keep it on the service side. A CommitMessage call should be issued to remove it from the service side.
If an error occures the message can be rereceived. This is QoS 1-level reliability from the receiver's side.
On error returns an uninitialized message.

@return */
PushedMessage OccasionallyConnectionClient::PeekMessage()
{
	return Receive("messages/peek");
}

/**
Removed the Message received by the PeekMessage call.
On error returns the http status code, otherwise 0.

@return */
int OccasionallyConnectionClient::CommitMessage()
{
	Response httpResponse = _restConnection->Post("messages/commit", "application/json", "");

	if(httpResponse.Code!= 204)
		return httpResponse.Code;

	return 0;
}

PushedMessage OccasionallyConnectionClient::Receive(const string& url)
{
	PushedMessage pushedMessage;

	Response httpResponse = _restConnection->Get(url);

	if(httpResponse.Body == "null")
	{
		pushedMessage.MessageId = -1;
		return pushedMessage;
	}

	if(httpResponse.Code!= 200)
	{
		pushedMessage.MessageId = -httpResponse.Code;
		return pushedMessage;
	}

	DynamicJsonBuffer jsonBufferResponse;
	JsonObject& respJson = jsonBufferResponse.parseObject((char *)httpResponse.Body.c_str());
	
	pushedMessage.MessageId = (int32_t)respJson["MessageId"];
	pushedMessage.SenderDeviceId = string(respJson["SenderDeviceId"].asString());
	pushedMessage.Payload = string(respJson["Payload"].asString());
	pushedMessage.Timestamp = (long unsigned int)respJson["Timestamp"];

	return pushedMessage;
}
}}

